codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt

```diff
@@ -114,7 +136,83 @@ class ServerProtocolTestGenerator(
         }
     }
 
-    fun render() {
+    fun render(writer: RustWriter) {
+        renderTestHelper(writer)
+
+        for (operation in operations) {
+            protocolGenerator.serverRenderOperation(writer, operation)
+            renderOperationTestCases(operation, writer)
+        }
+    }
+
+    /**
+     * Render a test helper module that help
+     *
+     * - generate dynamic builder for each handler, and
+     * - construct a tower service to exercise each test case.
+     */
+    private fun renderTestHelper(writer: RustWriter) {
+        // Create a tower service to perform protocol test
+        val operationNames = operations.map { it.toName() }
+        val operationRegistryName = "OperationRegistry"
+        val operationRegistryBuilderName = "${operationRegistryName}Builder"
+
+        val moduleMeta = RustMetadata(
+            additionalAttributes = listOf(
+                Attribute.Cfg("test"),
+                Attribute.Custom("allow(dead_code)"),
+            ),
+
+            visibility = Visibility.PUBCRATE,
+        )
+        writer.withModule("protocol_test_helper", moduleMeta) {
+            rustTemplate(
+                """
+                use #{Tower}::Service as _;
+
+                pub(crate) type F<Input, Output> = fn(Input) -> std::pin::Pin<Box<dyn std::future::Future<Output = Output> + Send>>;
+
+                type RegistryBuilder = crate::operation_registry::$operationRegistryBuilderName<#{Hyper}::Body, ${
+                operations.map {
+                    val (inputT, outputT) = operationInputOutputTypes[it]!!
+                    "F<$inputT, $outputT>, ()"
+                }.joinToString(", ")
+                }>;
+
+                fn create_operation_registry_builder() -> RegistryBuilder {
+                    crate::operation_registry::$operationRegistryBuilderName::default()
+                        ${operations.mapIndexed { idx, operationShape ->
+                    val (inputT, outputT) = operationInputOutputTypes[operationShape]!!
+                    ".${operationNames[idx]}((|_| Box::pin(async { todo!() })) as F<$inputT, $outputT> )"
+                }.joinToString("\n")}
+                }
+
+                pub(crate) async fn validate_request(
+                    http_request: #{Http}::request::Request<#{SmithyHttpServer}::body::Body>,
+                    f: &dyn Fn(RegistryBuilder) -> RegistryBuilder,
+                ) {
+                    let router: #{Router} = f(create_operation_registry_builder())
+                        .build()
+                        .expect("unable to build operation registry")
+                        .into();
+                    _ = router.into_make_service()
+                        .call(())
+                        .await
+                        .expect("unable to get a router")
+                        .call(http_request)
+                        .await
+                        .expect("unable to make an HTTP request");
```

Comment:
```suggestion
                    let mut router: #{Router} = f(create_operation_registry_builder())
                        .build()
                        .expect("unable to build operation registry")
                        .into();
                    _ = router
                        .call(http_request)
                        .await
                        .expect("unable to make an HTTP request");
```

There's no need to convert into a `MakeService` just to call the service once :).